src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/DPBasedJoinPlanOptimizer.java

```diff
@@ -0,0 +1,164 @@
+
+                        candidatePlans.add( PhysicalPlanFactory.createPlanWithJoin( plan_left,  plan_right) );
+
+                        if ( IdentifyTypeOfRequestUsedForReq.isBGPRequest( plan_left.getRootOperator() ) ){
+                            final LogicalOpBGPAdd newRoot = LogicalOpUtils.createBGPAddLopFromReq((BGPRequest) plan_left.getRootOperator());
+
+                            candidatePlans.add( PhysicalPlanFactory.createPlanWithIndexNLJ( newRoot, plan_right ) );
+                            candidatePlans.add( PhysicalPlanFactory.createPlanWithBindJoinFILTER( newRoot, plan_right ) );
+                            candidatePlans.add( PhysicalPlanFactory.createPlanWithBindJoinUNION( newRoot, plan_right ) );
+                            candidatePlans.add( PhysicalPlanFactory.createPlanWithBindJoinVALUES( newRoot, plan_right ) );
+                        }
+                        else if ( IdentifyTypeOfRequestUsedForReq.isTriplePatternRequest( plan_left.getRootOperator() ) ){
+                            final LogicalOpTPAdd newRoot = LogicalOpUtils.createTPAddLopFromReq((TriplePatternRequest) plan_left.getRootOperator());
+                            candidatePlans.add( PhysicalPlanFactory.createPlanWithIndexNLJ( newRoot, plan_right ) );
+
+                            final FederationMember fm = ( (LogicalOpRequest<?, ?>)plan_left.getRootOperator() ).getFederationMember();
+                            if ( fm instanceof SPARQLEndpoint ){
+                                candidatePlans.add( PhysicalPlanFactory.createPlanWithBindJoinFILTER( newRoot, plan_right ) );
+                                candidatePlans.add( PhysicalPlanFactory.createPlanWithBindJoinUNION( newRoot, plan_right ) );
+                                candidatePlans.add( PhysicalPlanFactory.createPlanWithBindJoinVALUES( newRoot, plan_right ) );
+                            }
+                        }
+                    }
+
+                    // Prune: only the best candidate plan is retained in optPlan.
+                    // TODO: Move the cost annotation out of this for-loop. For all plans of the same size, invoke the cost function once.
+                    final List<PhysicalPlanWithCost> candidatesWithCost = PhysicalPlanWithCostUtils.annotatePlansWithCost(ctxt.getCostModel(), candidatePlans);
+                    final PhysicalPlan planWithLowestCost = PhysicalPlanWithCostUtils.findPlanWithLowestCost(candidatesWithCost).getPlan();
+                    optPlan.add( plans, planWithLowestCost );
+                }
+            }
+            return optPlan.get( subplans );
+        }
+
+    }
+
+    // This method returns all subsets (with the given size) of the given superset.
+    public static <T> List<List<T>> getSubSet( final List<T> superset, final int n ){
+        final List<List<T>> result = new ArrayList<>();
+        if( n==0 ){
+            result.add( new ArrayList<>() );
+            return result;
+        }
```

Comment:
Actually, I would consider `n==0` to be an invalid argument (in fact, any `n<1`). You should throw an `IllegalArgumentException`.

Similarly, throw an `IllegalArgumentException` if `n>superset.size()`.

Additionally, if `n==superset.size()` you can immediately return `superset` without having to do anything else.